lark/lark_validator.py

```diff
@@ -0,0 +1,265 @@
+
+    def _ignore(self, stmt: Tree) -> None:
+        # Children: expansions
+        # - or -
+        # Children: token
+       exp_or_name = stmt.children[0]
+       if isinstance(exp_or_name, str):
+            self._ignore_names.append(exp_or_name)
+       else:
+            assert isinstance(exp_or_name, Tree)
+            t = exp_or_name
+            if t.data == 'expansions' and len(t.children) == 1:
+                t2 ,= t.children
+                if t2.data=='expansion':
+                    if len(t2.children) > 1:
+                        raise GrammarError("Bad %ignore - must have a Terminal or other value.")
+                    item ,= t2.children
+                    if item.data == 'value':
+                        item ,= item.children
+                        if isinstance(item, Token):
+                            # Keep terminal name, no need to create a new definition
+                            self._ignore_names.append(item.value)
+                            return
+                        if item.data == 'name':
+                            token ,= item.children
+                            if isinstance(token, Token) and token.type == "TOKEN":
+                                # Keep terminal name, no need to create a new definition
+                                self._ignore_names.append(token.value)
+                                return
+            name = '__IGNORE_%d'% len(self._ignore_names)
+            self._ignore_names.append(name)
+            self._definitions[name] = Definition(True, t, options=TOKEN_DEFAULT_PRIORITY)
+
+    def _literal(self, tree: Tree) -> None:
+        # Based on load_grammar.GrammarBuilder.literal_to_pattern().
+        assert tree.data == 'literal'
+        literal = tree.children[0]
+        assert isinstance(literal, Token)
+        v = literal.value
+        flag_start = max(v.rfind('/'), v.rfind('"'))+1
+        assert flag_start > 0
+        flags = v[flag_start:]
+        if literal.type == 'STRING' and '\n' in v:
+            raise GrammarError('You cannot put newlines in string literals')
+        if literal.type == 'REGEXP' and '\n' in v and 'x' not in flags:
+            raise GrammarError('You can only use newlines in regular expressions '
+                               'with the `x` (verbose) flag')
+        v = v[:flag_start]
+        assert v[0] == v[-1] and v[0] in '"/'
+        x = v[1:-1]
+        s = eval_escaping(x)
+        if s == "":
+            raise GrammarError("Empty terminals are not allowed (%s)" % literal)
+
+    def _load_grammar(self, tree: Tree) -> None:
+        for stmt in tree.children:
+            if stmt.data == 'declare':
+                self._declare(stmt)
+            elif stmt.data == 'extend':
+                self._extend(stmt)
+            elif stmt.data == 'ignore':
+                self._ignore(stmt)
+            elif stmt.data in ['import', 'multi_import']:
+                # TODO How can we process imports in the validator?
```

Comment:
imports always explicitly list the names being imported, so we can treat them like `%declare`, which should be decent enough, right?

Or maybe that produces problems with `%extend`, hm...

Reply:
Right, there's `%extend` and `%override`, at least.  I didn't feel up to recreating the logic of %import.

Reply:
If my comment of 2024-06-20 is acceptable, let's resolve this point.